src/Control/Carrier/NonDet/Church.hs

```diff
@@ -1,4 +1,14 @@
 {-# LANGUAGE DeriveTraversable, FlexibleInstances, MultiParamTypeClasses, RankNTypes, TypeOperators, UndecidableInstances #-}
+-- | Provides 'NonDetC', a carrier for 'NonDet' effects providing choice and failure.
+--
+-- It can be invoked with custom functions for choice, success, and failure ('runNonDet'), or it can delegate
+-- said operations to an 'Control.Applicative.Alternative' instance ('runNonDetA') or an underlying monoidal
+-- result ('runNonDetM'). Under the hood, it uses a Church-encoded structure and a binary tree to prevent
+-- the problems associated with a naïve list-based implementation.
+--
+-- This design is based on that detailed in Ralf Hinze's [Deriving Backtracking Monad Transformers](https://www.cs.ox.ac.uk/ralf.hinze/publications/#P12).
+--
+-- The carrier provided by "Control.Carrier.NonDet.Maybe" is similar, but can handle infinite search spaces at the cost of being able to return only one result.
```

Comment:
1. I don’t think we should cross-reference carriers in each carrier module, because that’s O(n²) documentation we have to write and maintain, and it’s a long way away from the things it’s talking about and thus prone to falling out of sync. Instead, we can cross-reference them in the effect modules in only O(n) docs.

2. This carrier can also handle infinite search spaces just fine; it’s all down to what you pass to `runNonDet`.